oauthenticator/globus.py

```diff
@@ -0,0 +1,123 @@
+        scope_string = (' ').join(scopes)
+        redirect_uri = self.authenticator.get_callback_url(self)
+        client = self.authenticator.globus_portal_client()
+        client.oauth2_start_flow(
+            redirect_uri,
+            requested_scopes=scope_string,
+            refresh_tokens=True)
+        self.log.info('globus redirect: %r', redirect_uri)
+        self.authorize_redirect(client)
+
+
+class GlobusOAuthenticator(OAuthenticator):
+
+    login_service = "Globus"
+    login_handler = GlobusLoginHandler
+
+    def get_globus_app_id(self):
+        """
+        Getting the custom configuration value globus_app_id.
+        globus_app_id is the UUID of the Globus app to authenticate
+        against
+        """
+        return self.config.GlobusOAuthenticator.globus_app_id
+
+    def get_globus_app_secret(self):
+        """
+        Getting the custom configuration value globus_app_secret.
+        globus_app_secret is the secret generated on the Globus app
+        website use with Confidential Globus Auth Apps.
+        """
+        return self.config.GlobusOAuthenticator.globus_app_secret
+
+    def globus_portal_client(self):
+        """
+        Create an Globus Auth ConfidentialAppAuthClient
+        Need a ConfidentialAppAuthClient because the NativeAppClient
+        would require user input, i.e. c/p the token from the website
+        somewhere, which we wont do... for now.
+        """
+        return globus_sdk.ConfidentialAppAuthClient(
+            self.get_globus_app_id(),
+            self.get_globus_app_secret())
+
+    @gen.coroutine
+    def authenticate(self, handler, data=None):
+        """
+        Overwritting the authenticate method with a Globus Auth specific one
+        """
+        code = handler.get_argument("code", False)
+        if not code:
+            raise web.HTTPError(400, "oauth callback made without a token")
+        # TODO: Configure the curl_httpclient for tornado
+        scopes = ['openid', 'profile', 'email',
+                  'urn:globus:auth:scope:auth.globus.org:view_identities']
+        scope_string = (' ').join(scopes)
+        redirect_uri = self.get_callback_url(self)
+        client = self.globus_portal_client()
+        client.oauth2_start_flow(
+            redirect_uri,
+            requested_scopes=scope_string,
+            refresh_tokens=True)
+        # Doing the code for token for id_token exchange
+        tokens = client.oauth2_exchange_code_for_tokens(code)
+        id_token = tokens.decode_id_token(client)
+        username = id_token.get('preferred_username')
+        if not username.endswith('@globusid.org'):
+            raise HTTPError(403, ("You are not signed in "
+                                  "to your {} account.".format("Globus ID")))
+        # Need to return a username without the "email" ending
+        username = username.split('@')[0]
+        # Checking the user exists... Paranoia!
+        try:
+            pwd.getpwnam(username)
```

Comment:
OAuthenticators don't generally require local users to exist. That's what the Local... variations are for. You can see examples of these in the other implementations here.